### ISSUE TITLE
NVDA hanging should never lock up the keyboard to allow for recovering via restarts

### ISSUE THREAD

**Issue by Neurrone** [nvaccess/nvda#10909]
### Steps to reproduce:

1. Do something that causes NVDA to reliably crash. For example, I can do this reliably by doing something that causes the command line to print a large amount of output. Its obviously not ideal that it crashes NVDA, but this is the most convenient way that I know of triggering this.

### Actual behavior:

Controlling the computer via keyboard is impossible, so the only way to recover is a forced shutdown via the physical power button and a restart, losing all work.

### Expected behavior:

Keyboard commands should still work, e.g windows+d, and navigating to NVDA and pressing enter.

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
Latest alpha
#### Windows version:
Windows 10 1809
#### Name and version of other software in use when reproducing the issue:
Anything that causes NVDA to hang, in my case it was a docker container unexpectedly causing me to restart my computer.

#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If addons are disabled, is your problem still occuring?
Irrelevant
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
No

**Comment by zahra21** [comment]
hi.
i confirm this problem.
i faced this issue several times,
as i remember, especially since the time that nvda migrated to espeak ng.
nvda 2017.1 was very great and i wish that i could use it forever!
even if i used laptop several hours and opened several times in my firefox, nvda was very responsive in compare to other versions!
when i installed 2017.2 for the first time, i opened firefox and firefox hanged and my laptop became completely unusable!
i fourced shutdown my laptop via physical power button.
in twelve days before now, i installed 2017.3,
and firefox becomes unresponsive and nvda and keyboard hang too!
i use firefox as my only web browser, i use windows xp 32bit service pack 3,
my laptop has 2.5 gb of ram,
2 cores of cpu with 65 nm! 1.73 ghz.
i dont use braille and now, since 2014, i only used espeak iven3 voice which is the best and most lovely voice in espeak!

**Comment by feerrenrut** [comment]
This should probably be thought of as an enhancement request rather than a bug report. The goal is to improve the user experience in the event of NVDA getting locked up. Unfortunately there are several ways this can happen, and likely require large changes to NVDA to improve the outcome.

**Comment by Neurrone** [comment]
@feerrenrut I disagree. Marking this as a UX enhancement trivializes the severity of the problem, and arguably NVDA hanging should not prevent any further keyboard usage. I understand that this won't be an easy problem to solve though, but marking this a bug would be more accurate.

**Comment by francipvb** [comment]
Hello,


What about migration to asyncio? This should solve a bunch of problems...

-- 
Cuando tus fuerzas terminan, las de mi Dios comienzan.

**Comment by leonardder** [comment]
I have looked into a migration to asyncio, but that also requires asyncio support for wx and that's not currently there.

**Comment by zersiax** [comment]
I have to agree with @Neurrone here. This is a bug, this is a showstopper, this can cause people to lose their jobs.
If this happens in the middle of an important consultation, or while you are working on something important that for whatever reason isn't autosaving, or a myriad of other factors, you will lose data. If you lose data on the job, that effectively means you lose money. If you lose money because of a screen reader, that screen reader cannot be used in a work setting. It really is as simple as that.
I appreciate that this will require large changes on NVDA's side. I also think that this has happened often enough over the last years that these changes are warranted, especially after the python3 migration. If there are ways this can be mittigated, I feel that needs to be looked into.
Not hearing "subject: blah"  but rather " blah"  for an email subject is a UX enhancement. This is a critical bug to those it happens to.

**Comment by leonardder** [comment]
Related to #10247, #6463, #6291

**Comment by leonardder** [comment]
I tend to agree with @zersiax here. The problem described here shouldn't be treated as an enhancement, but rather as a p2 issue. See also triage of issues mentioned above.

**Comment by Adriani90** [comment]
@Neurrone are you still having this issue with NVDA 2024.4.1?

Anyway, I think migrating to asyncio makes definitely sense, and it is an official library in Python 3.7 and higher. It is also supported by WX Python. I found also these libraries which might be useful:
https://github.com/sirk390/wxasync
and
https://pypi.org/project/wxasync/

I think this might improve performance in lots of use cases if event loops are not blicking the main thread so maybe someone can look into such a migration. It seems it can be imported already in python environments:


cc: @seanbudd, @michaelDCurran maybe you can give some insights of the potential this would have on NVDA? Is there any plan to migrate in the near future?

**Comment by Neurrone** [comment]
This is definitely still a problem. I don't see this often thankfully but when it does, it is crippling.
